napari/_qt/widgets/qt_dims_sorter.py

```diff
@@ -46,55 +30,85 @@ def _array_in_range(arr: np.ndarray, low: int, high: int) -> bool:
             return
-        axes = [a.axis for a in axes_list]
+        axes = [a.axis for a in axis_list]
         ax_to_existing_position = {a: ix for ix, a in enumerate(axes)}
         move_list = np.asarray(
             [(ax_to_existing_position[order[i]], i) for i in range(len(order))]
         )
         for src, dst in move_list:
-            axes_list.move(src, dst)
+            axis_list.move(src, dst)
             move_list[_array_in_range(move_list[:, 0], dst, src)] += 1
         # remove the elements from the back if order has changed length
-        while len(axes_list) > len(order):
-            axes_list.pop()
+        while len(axis_list) > len(order):
+            axis_list.pop()
 
 
 class QtDimsSorter(QWidget):
-    """
+    """Qt widget for dimension / axis reordering and locking.
+
     Modified from:
     https://github.com/jni/zarpaint/blob/main/zarpaint/_dims_chooser.py
+
+    Parameters
+    ----------
+    viewer : napari.Viewer
+        Main napari viewer instance.
+    parent : QWidget
+        QWidget that holds this widget. A QtDimsSorter instances will
+        disconnect all callbacks upon closing of it's parent.
+
+    Attributes
+    ----------
+    dims : napari.components.Dims
+        Dimensions object of the current viewer, modeling slicing and displaying.
+    axis_list : napari._qt.containers.qt_axis_model.AxisList
+        Selectable evented list representing the viewer axes.
     """
 
-    def __init__(self, viewer: 'Viewer', parent=None) -> None:
+    def __init__(self, dims: Dims, parent: QWidget) -> None:
         super().__init__(parent=parent)
-        dims = viewer.dims
-        root = SelectableEventedList(
-            [AxisModel(dims, dims.order[i]) for i in range(dims.ndim)]
-        )
-        root.events.reordered.connect(
-            lambda event: set_dims_order(dims, event.value)
-        )
-        dims.events.order.connect(
-            lambda event: move_indices(root, event.value)
-        )
-        view = QtListView(root)
-        view.setSizeAdjustPolicy(QtListView.AdjustToContents)
+        self.dims = dims
+        self.axis_list = AxisList.from_dims(self.dims)
 
-        self.axes_list = root
+        view = QtListView(self.axis_list)
+        view.setSizeAdjustPolicy(QtListView.AdjustToContents)
 
         layout = QGridLayout()
         self.setLayout(layout)
 
         widget_tooltip = QtToolTipLabel(self)
         widget_tooltip.setObjectName('help_label')
-        widget_tooltip.setToolTip(trans._('Drag dimensions to reorder.'))
+        widget_tooltip.setToolTip(
+            trans._(
+                'Drag dimensions to reorder, click lock icon to lock dimension in place.'
+            )
+        )
 
         widget_title = QLabel(trans._('Dims. Ordering'), self)
 
         self.layout().addWidget(widget_title, 0, 0)
         self.layout().addWidget(widget_tooltip, 0, 1)
-        self.layout().addWidget(view, 1, 0)
+        self.layout().addWidget(view, 1, 0, 1, 2)
+
```

Comment:
```suggestion
        layout.setContentsMargins(0, 0, 0, 0)
```
The current widget looks rather bloated, compared to for example the Grid button pop-up widget.

Honestly, I'd like the QListView to just show all the dims instead of the scroll bar, to make it easier to interact with, but I couldn't figure out the size policy.
Maybe Qt 🧙 @dalthviz can help?

Reply:
Hi! One way to show all the dims could be to remove the line that sets a `sizeAdjustPolicy` for the `QtListView` instance at:

https://github.com/napari/napari/blob/558ffa6fd561cef30146eabe5a5e136bd6b47eb6/napari/_qt/widgets/qt_dims_sorter.py#L79

However, that causes the popup to be quite big:

![imagen](https://github.com/napari/napari/assets/16781833/4dbfb7b5-0e65-4933-8877-e04b3faf7386)


To prevent that you could set a style rule with `max-width` and `max-height`. Something like:

```css
QtDimsSorter > QtListView {
  max-height: 100px;
  max-width: 150px; 
}
```
With that then you can get something like:

![imagen](https://github.com/napari/napari/assets/16781833/a526571a-8a5d-4c36-b8d2-2a254d023ae5)

So you at least can define the point when a scrollbar appears. For example, if I set `50px` as `max-height` I'm seeing something like:

![imagen](https://github.com/napari/napari/assets/16781833/df187fe2-de7b-47d2-82cf-e2c39e1fb3fc)

Reply:
Thanks! Yeah I got the first one (huge pop-up) but forgot that using qss could help here. Your 2nd image above is exactly what I had in mind: show 3 and then scrollbar if 3+ (with just 2dims it's the degenerate case so I don't think it matters if there's a blank space below)

Reply:
Actually, checking a little bit more this I think we could prevent the blank space for the 2 dims case by instead of fully removing the `SetSizeAdjustPolicy` line, doing the call only for <=2 dims cases. If I do something like:

```python
    if len(self.axis_list) <= 2:
        view.setSizeAdjustPolicy(QtListView.AdjustToContents)
```
Then we can get the following behavior:

![popup_dim_sorter](https://github.com/napari/napari/assets/16781833/7ffdca5c-618b-4071-847f-f878a2476bd6)